### PR TITLE
ziga/add new obscuro gateway endpoints

### PR DESCRIPTION
**Pull request by zkokelj**
### Why this change is needed

Please provide a description and a link to the underlying ticket

### What changes were made as part of this PR

Please provide a high level list of the changes made

### PR checks pre-merging

Please indicate below by ticking the checkbox that you have read and performed the required
[PR checks](https://github.com/obscuronet/obscuro-internal/blob/main/dev-ops-docs/dev-pr-checks.md)

- [ ] PR checks reviewed and performed

### CODE REVIEW

File: go/rpc/viewing_key.go [hunk 1]
```diff
@@ -20,7 +20,8 @@ type ViewingKey struct {
 	Account    *common.Address   // Account address that this private key is bound to
 	PrivateKey *ecies.PrivateKey // private viewing key
 	PublicKey  []byte            // public viewing key in bytes to share with enclave
-	SignedKey  []byte            // public viewing key signed by the Account's private key
+	// todo (@ziga) - remove signed key, because we wont sign key, but following string "Register $UserId for $Acct"
```
Comment (tudor-malene) [inline go/rpc/viewing_key.go:23]: technically, before we signed over a text as well. The text was "vk$key"

File: tools/walletextension/api/routes.go [hunk 2]
```diff
@@ -216,3 +241,124 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+
+	// save UserID and PrivateKey to the database
+	vk := &rpc.ViewingKey{
+		Account:    nil,
+		PrivateKey: viewingPrivateKeyEcies,
+		PublicKey:  viewingPublicKeyBytes,
+		SignedKey:  nil, // we await a signature from the user before we can set up the EncRPCClient
+	}
+
+	err = walletExt.Storage.SaveUserVK(userID.Hex(), vk, "")
```
Comment (tudor-malene) [inline tools/walletextension/api/routes.go:354]: why store the hex. You can store the binary form.

File: tools/walletextension/api/routes.go [hunk 3]
```diff
@@ -216,3 +241,124 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+	_, err := userConn.ReadRequest()
+	if err != nil {
+		return
+	}
+
+	// generate new key-pair
+	viewingKeyPrivate, err := crypto.GenerateKey()
+	viewingPrivateKeyEcies := ecies.ImportECDSA(viewingKeyPrivate)
+	if err != nil {
+		userConn.HandleError(fmt.Sprintf("could not generate new keypair: %s", err))
```
Comment (tudor-malene) [inline tools/walletextension/api/routes.go:337]: these errors should be returned as "Internal error" to the user, and logged with this detailed message.

File: tools/walletextension/api/routes.go [hunk 4]
```diff
@@ -216,3 +241,124 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
 }
+
+func authenticateRequestHandler(walletExt *walletextension.WalletExtension, userConn userconn.UserConn) {
+	// check if the text is well-formed and extract userID and address
+	body, err := userConn.ReadRequest()
+	if err != nil {
+		return
+	}
+
+	var reqJSONMap map[string]string
```
Comment (tudor-malene) [inline tools/walletextension/api/routes.go:252]: this needs some comments.
I don't quite follow what's going on here.

Based on the design there should be 2 fields in the json:
- The text that was signed
- the signature

What I imagined is that:
1. parse the text and extract the userId and the account.
2. from the db, fetch the VK entry for the userId
3. Check the signature of the account over text
4. Register the account against the VK

File: tools/walletextension/storage/sqlite.go [hunk 5]
```diff
@@ -22,11 +22,15 @@ func NewSqliteDatabase(dbName string) (*SqliteDatabase, error) {
 		return nil, err
 	}
 
+	// todo (@ziga) - should we use binary format also for signature and text
 	_, err = db.Exec(`CREATE TABLE IF NOT EXISTS viewingkeys (
```
Comment (tudor-malene) [inline tools/walletextension/storage/sqlite.go:26]: I think we should have 2 tables:
1. Users (user_id, private_key)
2. Accounts(user_id, account_address, signature)

The signature is over a text in the right format, which you can recreate from the user_id, and account_address

**Review by tudor-malene — COMMENTED** [review]
had a quick look and left a few comments.

File: tools/walletextension/api/routes.go [hunk 6]
```diff
@@ -86,6 +101,16 @@ func NewWSRoutes(walletExt *walletextension.WalletExtension) []Route {
 			Name: common.PathSubmitViewingKey,
 			Func: wsHandler(walletExt, submitViewingKeyRequestHandler),
 		},
+
```
Comment (otherview) [inline tools/walletextension/api/routes.go:104]: Will these paths be accessed over WS ?

File: tools/walletextension/api/routes.go [hunk 7]
```diff
@@ -216,3 +241,143 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
 		return
 	}
 }
+
+func authenticateRequestHandler(walletExt *walletextension.WalletExtension, userConn userconn.UserConn) {
+	// check if the text is well-formed and extract signature and message
+	body, err := userConn.ReadRequest()
+	if err != nil {
+		return
```
Comment (otherview) [inline tools/walletextension/api/routes.go:249]: errs should be logged if not returned

File: tools/walletextension/api/routes.go [hunk 8]
```diff
@@ -216,3 +241,143 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+	}
+
+	var reqJSONMap map[string]string
+	err = json.Unmarshal(body, &reqJSONMap)
+	if err != nil {
+		userConn.HandleError(fmt.Sprintf("could not unmarshal viewing key and signature from client to JSON: %s", err))
```
Comment (otherview) [inline tools/walletextension/api/routes.go:255]:
```suggestion
		userConn.HandleError(fmt.Errorf("could not unmarshal viewing key and signature from client to JSON: %w", err).Error())
```

File: tools/walletextension/api/routes.go [hunk 9]
```diff
@@ -216,3 +241,143 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
...
+		return
+	}
+
+	signature, err := hex.DecodeString(reqJSONMap[common.JSONKeySignature][2:])
```
Comment (otherview) [inline tools/walletextension/api/routes.go:259]: That `[2:]` is a magic number, helpful to have an explanation in a comment

**Comment by zkokelj** [comment]
After the discussion with @otherview I decided to split this big PR to multiple smaller PRs.
All the comments above were taken into account when making changes to the code in new PRs.

**Pull request closed without merging**